text/0002-signature-based-completions.md

```diff
@@ -0,0 +1,216 @@
+- Feature Name: signature-based-completions
+- Start Date: 2020-07-25
+- RFC PR: [nushell/rfcs#0002](https://github.com/nushell/rfcs/pull/2)
+- Nushell Issue: [nushell/nushell#0000](https://github.com/nushell/nushell/issues/0000)
+
+# Summary
+
+Add a completion engine to nushell that can determine how to complete an item at the cursor by
+parsing the line, extracting a signature, and using the information in that signature to determine
+what kind of thing needs to be completed.
+
+# Motivation
+
+We currently use rustyline's completer to do our completions. This completer is unfortunately
+limited, as it focuses on filenames only. We've tried to expand this with command names, but it
+scatters completion code all over the codebase.
+
+We already have incredibly rich information in signatures for commands, and intend to eventually use
+this to describe external commands. By building our completions around signatures, we can take
+advantage of highly contextual information to make nushell's completion system highly contextual,
+powerful, and delightful to use.
+
+# Guide-level explanation
+
+Completion is the act of responding to a user request to find a list of recommendations that fit
+some incomplete information that exists at their cursor. For example, if the user presses tab at the
+end of this line:
+
+    > ls dir/
+
+all path entries under the subdirectory `dir` could be offered as suggestions. Similarly, for
+
+    > ls dir/a
+
+we'd expect all path entries under `dir` that start with `a`. It's also possible for the cursor to
+be in the middle of a line, too. For
+
+    > ls di<TAB>/a
+
+Here we can show all directories whose name starts with "di". If the user chooses one of the
+suggestions, completion would complete up to the `/`. Note, there are alternative approaches to how
+completion could work here, but this RFC will assume the aforementioned, as it is the simplest. It
+should be easy to switch to a smarter completer in the future though.
+
+Note that this is different from "hinting", which looks at the line before the cursor and provides a
+suggestion for how to complete the rest of the line. The major difference is that hinting will
+replace the entire line, has less context, and is typically based on command history. This RFC does
+not address hinting.
```

Comment:
What about partial completion that uses hinting and allows you to tab through the tokens up to the point that you want a change? This would be invaluable for nu since our pipeline command can get very long.

Reply:
Do you mean that if we have this line:

```
cmd --thing
```

And the hinted line were

```
cmd --thing "my value" "some positional argument" 5
```

That you could work through the parts of the hint in this order:

1. `cmd --thing "my value"`
2. `cmd --thing "my value" "some positional argument"`
3. `cmd --thing "my value" "some positional argument" 5`

If so, agree that's super useful, but still outside the context of this RFC. Mind you, the work we do in this RFC will help with that, but hinting and completions are different beasts, so I'm trying to simplify this RFC by only considering the latter.

Reply:
ya, I think you get what I'm saying. we have some like `cmd | cmd | cmd | cmd | cmd | cmd | cmd` and i want to change the 5th cmd without retyping the whole thing.

Reply:
Yeah, I think we're on the same page 👍 

So, definitely, this is important, but not for consideration in this PR. "Hinting" is all about completing the line based on historical data. "Completion" is all about taking the context of where the cursor is and providing values that could fill that position. Do you think we could use better naming? Maybe "Contextual completion" vs "Line completion"?

Just to help show how I think they're sufficiently different to keep separate, take the cursor being located at the `|` in this example:

```
command --my-flag=| some more arguments
```

Hinting would just consider the entirety of the line as a prefix against the history, so it would try to find something in history that has `command --my-flag= some more arguments` as a prefix. Completion needs to understand that the cursor is in the value position for the flag named  `my-flag`, and also understand what kind of value `my-flag` is expecting. From that information, a list of possible suggestions are given, which will replace that portion of the line. There's some similarity, but I think it's fairly minimal.

Another thing to consider is your example:

```
cmd<TAB>
```

Should that take the next piece in the hint? Should it suggest any internal/external/alias whose name starts with cmd?

All in all, hinting will be important to improve on (probably won't need an RFC). I expect the work I do for completions will not have any negative impact on improving hinting.

Reply:
I tend to talk authoritatively even when I don't know what I'm talking about so don't take this as my way or the high way demands. I'm just throwing this out.

`cmd<TAB>` should (in this order):

1. If hints were supported, it should try to complete the next token in the hint string (history line).
2. Look for internal commands to complete with.
3. Look for subdirectories off the cwd, only if the spacially delimited token in front of it is a folder directive like `ls`, `cd`, `mv`. In this case this step is skipped because there is no folder directive.
4. Look for external commands to complete with i.e. executables in the path.
5. Not try to complete flags or sub commands.


`cmd <TAB>` should (in order):

1. Look for subcommands to complete with.
2. Look for flags to complete with.


`cmd -<TAB>` should:

1. Look at flags to complete with.

Reply:
👍 I've never used a shell that was configured such that (1) happens, so it's not something I'm familiar with. My own preference is actually for completions and hints to be "activated" separately, generally the tab key for completions, and the right arrow key for hinting (possibly with some modifier key to only partially complete).

Raising that as a potential option though, means I should at least survey things a little more to understand how everyone else does it. I'll explore this a little further and update the RFC with my findings 🙂

Reply:
So, I'll update the RFC calling out that hinting vs completion, as I've described it, are different in other shells too (findings below). The tl;dr is that you typically complete with TAB and you hint with Up/Down, sometimes with Alt to do a partial completion.

Now, that doesn't mean history can't be used as a completer, but it's not completing the line, just the contextual item that completion would normally complete. For example, if I have an integer flag, I could use values from historical lines using the same command to offer as completion suggestions.

# Findings

## Bash

```
man bash
```

> `reverse-search-history (C-r)`
>     Search backward starting at the current line and moving `up' through the history as necessary.  This is an incremental search.
> `forward-search-history (C-s)`
>     Search forward starting at the current line and moving `down' through the history as necessary.  This is an incremental search.

## Zsh

```
man zshall
```

> A  history  mechanism  for retrieving previously typed lines (most simply with the Up or Down arrow keys) is available

## Fish

https://fishshell.com/docs/current/#history-search

Up/down will do a full history completion with the line as a prefix
Up/down + Alt will do a partial completion with the line as a prefix

## Powershell

https://www.powershellmagazine.com/2013/07/11/pstip-using-tab-completion-against-the-command-history/

> Each time you execute a command in Windows PowerShell, the command is remembered and can be fetched by using the Get-History cmdlet or by using the the up/down arrow keys.

That being said, if you put a `#` at the beginning, then you can use TAB apparently:

> For example, to recall the first command containing the string ‘dir’, type the ‘#’ sign followed by the pattern of the command you want to find, and then press the Tab key:

I don't have a Windows machine to verify this though.

Reply:
I'm going to resolve this comment, @fdncred. I think I've addressed your comment, showing the difference between hinting and completion, but reopen if you think I haven't.

Reply:
> For example, to recall the first command containing the string ‘dir’, type the ‘#’ sign followed by the pattern of the command you want to find, and then press the Tab key:

This works and I didn't know it. Thanks! :)